califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
+        if (!fSimulation)
+        {
+
+            if (cryEnergy >= fGammaClusterThreshold && cryEnergy <= fSaturation)
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId > fTotalCrystals && cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId <= fTotalCrystals && cryEnergy > fSaturation)
+                saturatedCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+
+        /* -------- Simulation data ------- */
+        else
+        {
+
+            if (cryEnergy >= fGammaClusterThreshold && cryEnergy < fProtonClusterThreshold)
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+    }
+
+    /* ----- Remove duplicate entries in gamma clusters ----- */
+    for (int k = 0; k < gammaCandidatesVec.size(); k++)
+    {
+        for (int s = 0; s < gammaCandidatesVec.size(); s++)
+        {
+
+            if (gammaCandidatesVec.at(k)->GetCrystalId() == (gammaCandidatesVec.at(s)->GetCrystalId() - fTotalCrystals))
+                gammaCandidatesVec.erase(gammaCandidatesVec.begin() + s);
+        }
+    }
+
+    // Sort all vectors by energy
+    std::sort(gammaCandidatesVec.begin(), gammaCandidatesVec.end(), compareByEnergy);
+    std::sort(protonCandidatesVec.begin(), protonCandidatesVec.end(), compareByEnergy);
+
+    TVector3 mother_angles, angles;
+    Double_t fRandTheta, fRandPhi;
+
+    while (protonCandidatesVec.size())
+    {
+
+        Int_t motherId = protonCandidatesVec.at(0)->GetCrystalId();
+
+        califa_candidate cluster = { motherId, vector<Int_t>(), 0.0, 0.0, 0.0, 0.0, 0.0 };
+
+        if (motherId <= fTotalCrystals)
+            mother_angles = fCalifaGeo->GetAngles(motherId) - fCalifatoTargetPos;
+
+        else
+            mother_angles = fCalifaGeo->GetAngles(motherId - fTotalCrystals) - fCalifatoTargetPos;
+
+        if (fRand)
+        {
+
+            if (motherId > fTotalCrystals)
+                fAngularDistributions[protonCandidatesVec.at(0)->GetCrystalId() - 1 - fTotalCrystals]->GetRandom2(
+                    fRandPhi, fRandTheta);
+
+            else
+                fAngularDistributions[protonCandidatesVec.at(0)->GetCrystalId() - 1]->GetRandom2(fRandPhi, fRandTheta);
+
+            cluster.theta = TMath::DegToRad() * fRandTheta;
+            cluster.phi = TMath::DegToRad() * fRandPhi;
+        }
+
+        else
+        {
+
+            cluster.theta = mother_angles.Theta();
+            cluster.phi = mother_angles.Phi();
+        }
+
+        cluster.time = protonCandidatesVec.at(0)->GetTime();
```

Comment:
I think it's okay to assume the "time" from the highest-energy deposited crystal. But then, I think it's important to check the time differences from other crystals when applying the clustering to eliminate the background contributions.

Reply:
Requiring closeness in space and time instead of just requiring closeness in space is something I have thought a bit on. In general, the higher the energy, the better the time resolution. (Mostly not Heisenberg, I think.) So two crystals which were both directly hit by a proton should have a time difference of perhaps 100ns, while some crystal which only saw 100keV may be off by 1us perhaps. 
One way to go about this would be to have an energy dependent filter which selects the events which could have come at the time LOS recorded and rejects all the stuff which got timestitched but where the WRTS difference to los is to large for the crystal energy. Then just cluster the rest. (This is assuming that LOS had hit multiplicity of 1 and you don't care about the other events.)

Enough rambling for me, good night, and sorry for all the notification mails.

Reply:
Already answered.